content/blog/2022-04-26-april-22-community-gems.md

```diff
@@ -0,0 +1,124 @@
+  updates, working with multiple datasets, using DVC stages, and more.
+descriptionLong: >
+  A roundup of technical Q&A's from the DVC and CML community. This month: CML
+  updates, working with multiple datasets, using DVC stages, and more.
+picture: 2022-04-326/apr-comm-gems.png
+author: milecia_mcgregor
+commentsUrl:
+tags:
+  - CML
+  - DVC Stages
+  - DVC Remotes
+  - Community
+---
+
+### [When I run `dvc repro` on a stage, does it automatically push any outputs to my remote?](https://discord.com/channels/485586884165107732/563406153334128681/953616587523498025)
+
+Great question from @tina_rey!
+
+The `dvc repro` command doesn't automatically push any outputs or data to your
+remote. The outputs are stored in the cache until you run `dvc push`, which then
+pushes them from your cache to your remote.
+
+### [Is `dvc dag` based on `deps` and `outs`, so that a stage that depends on the output of another stage will always be executed after the former has finished?](https://discord.com/channels/485586884165107732/563406153334128681/956113493155799070)
+
+This is a good question from @johnysku!
+
+That is correct! If the pipelines are independent or the stages are independent,
+they may run in any order. Without explicit dependency linkage, stages could be
+executed in an unexpected order.
+
+One way around this is by outputing some kind of placeholder file when all a
+stage is finished and make that the dependency that triggers the next stage.
+
+<!-- TODO: add stage examples and dvc dag outputs -->
+
+### [If I want to use the `foreach` utility in `dvc repro`, is there a way I can use glob patterns to create the list DVC needs to iterate over?](https://discord.com/channels/485586884165107732/563406153334128681/956241424150577233)
+
+Another interesting question from @copah!
+
+If you have `mystage` which uses `foreach`, you can do `dvc repro` to `mystage`
+to iterate over every `mystage` stage.
+
+<!-- TODO: add stage examples for the foreach -->
+
+### [How does DVC handle files that have been deleted from remote storage?](https://discord.com/channels/485586884165107732/563406153334128681/956254582676258866)
+
+Really good question from @Meme Philosopher!
+
+DVC will fail when you try to pull files that have been deleted from the remote
+and notify you that those files are missing in remote storage.
+
+### [Can I separate CML running from GitHub actions VM to work with GCP or AWS so training and testing are in these cloud environments?](https://discord.com/channels/485586884165107732/728693131557732403/954316332457947169)
+
+Thanks for the question @Atsu!
+
+This is supported out-of-the-box! Here's how it works:
+
+1. GitHub action launches "self-hosted" GCP or AWS using
+   `cml runner --reuse --labels=cml`
+2. GitHub action runs the rest of the workflow on the "self-hosted" runner using
+   `runs-on: [self-hosted, cml]` and `timeout-minutes: 4320`
+3. If GitHub action is about to timeout, CML will restart the workflow.
```

Comment:
```suggestion
1. Within Github Actions, CML launches a [self-hosted runner](https://cml.dev/doc/self-hosted-runners) on GCP or AWS using
   `cml runner --labels=cml --cloud=gcp`/`--cloud=aws`
2. GitHub Actions runs the rest of the workflow on the self-hosted runner using
   `runs-on: [self-hosted, cml]` and the maximum allowable `timeout-minutes: 4320`
3. If GitHub Actions is about to timeout, CML will restart the workflow,
   so make sure your code regularly caches and restores data if it's expected to
   take >3 days to run.
```

Reply:
Thanks for this! That actually helped me understand how CML works even better.